### PR TITLE
Util

### PR DESCRIPTION
**Pull request by mortonjt**
This contains functions for

- Parsing filenames for the sample config
- Parsing sample sheets

This is required for #42 to completely work

### CODE REVIEW

File: oecophylla/util/parse.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        if s in samples:
+            samples_dict[s]['filter_db'] = db_fp
+        elif 'filter_db' in samples_dict[s]:
+            continue
+        else:
+            samples_dict[s]['filter_db'] = None
+
+    return(samples_dict)
+
+# Option 2: read samples from manifest
```
Comment (tanaes) [inline oecophylla/util/parse.py]: from sample sheet. ;P

File: oecophylla/util/parse.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+# Option 1 : parse samples from directory
+
+def parse_filenames_to_df(filepaths, pattern = None, column_names = None):
+    """ Generates sample sheet from filenames using a regex pattern.
+
+    Parameters
+    ----------
+    filepaths: list of str
+        List of file paths to parse
+    pattern: str
```
Comment (wasade) [inline oecophylla/util/parse.py]: ", optional"

File: oecophylla/util/parse.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        Regex pattern to recognize the filenames, to load into the
+        sample sheet.
+        default: '^((.+?)_(S\d+)_(L\d+)_(R[12])_(\d+)\.(.+))$'
+    column_names: list
```
Comment (wasade) [inline oecophylla/util/parse.py]: ", optional"

File: oecophylla/util/parse.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+def get_sample_reads_df_from_sample_sheet(ss_df, seq_dir,
+                                          sample_col='Description',
+                                          prefix_col='Sample_ID'):
+    sample_reads_dict = {}
+
+    samples = list(set(ss_df[sample_col]))
+
+    fps = os.listdir(seq_dir)
+
+    files_df = parse_ilm_fps_to_df(fps)
```
Comment (tanaes) [inline oecophylla/util/parse.py]: is `parse_ilm_fps_to_df` a valid def?

File: oecophylla/util/parse.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    column_names: list
+        Column names for the generated sample sheet.  Note that this
+        is assumed to be in the same ordering as the regex pattern.
+        default: ['File', 'Sample', 'Index', 'Lane',
+                  'Read', 'Run', 'Extension']
+
+
+    """
+    if pattern is None:
+        pattern = '^((.+?)_(S\d+)_(L\d+)_(R[12])_(\d+)\.(.+))$'
```
Comment (wasade) [inline oecophylla/util/parse.py]: should we assume an extension? `fastq.gz` would be great

File: oecophylla/util/parse.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    for f in filepaths:
+        m = p.match(f)
+        if m:
+            df = df.append(dict(zip(column_names, m.groups())),
+                           ignore_index = True)
+    return df
+
+
+def get_sample_reads_df(df, seq_dir):
+    """
```
Comment (wasade) [inline oecophylla/util/parse.py]: a bit incomplete..

File: oecophylla/util/parse.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    Parameters
+    ----------
+    df : pd.DataFrame
+
+    seq_dir : str
+
+    """
+    sample_reads_dict = {}
+
+    samples = list(set(df['Sample']))
```
Comment (wasade) [inline oecophylla/util/parse.py]: `samples = tuple(df['Sample'].unique())`

File: oecophylla/util/parse.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        is assumed to be in the same ordering as the regex pattern.
+        default: ['File', 'Sample', 'Index', 'Lane',
+                  'Read', 'Run', 'Extension']
+
+
+    """
+    if pattern is None:
+        pattern = '^((.+?)_(S\d+)_(L\d+)_(R[12])_(\d+)\.(.+))$'
+    if column_names is None:
+        column_names = ['File', 'Sample', 'Index', 'Lane',
```
Comment (wasade) [inline oecophylla/util/parse.py:37]: it looks like downstream code (e.g., `get_sample_reads_df` assumes at least some of these columns exist. Is there a usecase for the column names being parameterized (same q for the pattern)?

File: oecophylla/util/parse.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    seq_dir : str
+
+    """
+    sample_reads_dict = {}
+
+    samples = list(set(df['Sample']))
+
+    for s in samples:
+        f_fps = sorted(list(df.loc[(df['Sample'] == s) & (df['Read'] == 'R1'),'File'].values))
```
Comment (wasade) [inline oecophylla/util/parse.py]: might be a bit more readable with a `for sample, sample_grp in samples.groupby('Sample'): ...`

File: oecophylla/util/parse.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        r_fps = sorted(list(df.loc[(df['Sample'] == s) & (df['Read'] == 'R2'),'File'].values))
+
+        sample_reads_dict[s] = {'forward': [os.path.join(seq_dir, x) for x in f_fps],
```
Comment (wasade) [inline oecophylla/util/parse.py]: can we assert these files exist and raise an IOError if not?

File: oecophylla/util/parse.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                                'reverse': [os.path.join(seq_dir, x) for x in r_fps]}
+
+    return(sample_reads_dict)
+
+def get_sample_paths(seq_dir):
+    """
```
Comment (wasade) [inline oecophylla/util/parse.py]: ...also incomplete. i think either use numpydoc or dont

File: oecophylla/util/parse.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    files_df = parse_ilm_fps_to_df(fps)
+    sample_reads_dict = get_sample_reads_df(files_df, seq_dir)
+
+    return(sample_reads_dict)
+
+
+def add_filter_db(sample_fp_dict, db_fp, samples = None):
+
+    if samples is None:
+        samples = sample_fp_dict.keys()
```
Comment (wasade) [inline oecophylla/util/parse.py]: `samples = set(sample_fp_dict)`

File: oecophylla/util/parse.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        else:
+            samples_dict[s]['filter_db'] = None
+
+    return(samples_dict)
+
+# Option 2: read samples from manifest
+def read_sample_sheet(f, sep='\t'):
+    data = False
+    data_lines = ''
+    for line in f:
```
Comment (wasade) [inline oecophylla/util/parse.py]: a comment showing an example of what this format would be useful

File: oecophylla/util/parse.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+    samples = list(set(ss_df[sample_col]))
+
+    fps = os.listdir(seq_dir)
+
+    files_df = parse_ilm_fps_to_df(fps)
+
+    for s in samples:
+        # get the subset of sample sheet rows for that sample
+        sample_rows = ss_df.loc[ss_df[sample_col] == s,]
```
Comment (wasade) [inline oecophylla/util/parse.py]: `groupby`?

File: oecophylla/util/parse.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+                                      (files_df['Read'] == 'R1'), 'File'].values)
+
+            r_fps.extend(files_df.loc[(files_df['Sample'] == row['Sample_ID']) &
+                                      (files_df['Lane'] == lane) &
+                                      (files_df['Read'] == 'R2'), 'File'].values)
+
+        f_fps = sorted(f_fps)
+        r_fps = sorted(r_fps)
+
+        sample_reads_dict[s] = {'forward': [os.path.join(seq_dir, x) for x in f_fps],
```
Comment (wasade) [inline oecophylla/util/parse.py]: same q on filepath existence

File: oecophylla/util/tests/test_parse.py [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+                'Run':['001', '001', '001', '001', '001', '001',
+                       '001', '001', '001', '001', '001', '001'],
+                'Extension': ['fastq.gz']*12
+            }
+        )
+        exp_df = exp_df.reindex(columns= ['File', 'Sample', 'Index', 'Lane',
+                                          'Read', 'Run', 'Extension'])
+
+        pdt.assert_frame_equal(df, exp_df)
+
```
Comment (wasade) [inline oecophylla/util/tests/test_parse.py]: should there be more tests?

File: oecophylla/util/parse.py (hunk 9 above)
Comment (mortonjt) [reply oecophylla/util/parse.py]: I like it!

File: oecophylla/util/parse.py (hunk 8 above)
Comment (mortonjt) [reply oecophylla/util/parse.py:37]: Done

File: oecophylla/util/parse.py (hunk 14 above)
Comment (mortonjt) [reply oecophylla/util/parse.py]: Done

File: oecophylla/util/parse.py (hunk 13 above)
Comment (mortonjt) [reply oecophylla/util/parse.py]: The file is too big for the docs.  Here is an example [file](https://github.com/biocore/oecophylla/blob/master/test_data/example_sample_sheet.txt)

File: oecophylla/util/parse.py (hunk 10 above)
Comment (mortonjt) [reply oecophylla/util/parse.py]: I don't know how to do this in this particular scenario.  Raising issue.

**Comment by mortonjt** [comment]
This is ready for review / merge @wasade @tanaes @tkosciol

File: scripts/oecophylla.py [hunk 17]
```diff
@@ -4,11 +4,89 @@
...
+@click.option('--sample-sheet', required=True, type=click.STRING,
+              callback=_arg_split,
+              help='Sample sheets used to demultiplex the illumina run.')
+@click.option('--params', type=click.PATH, required=True,
+              help='Specify parameters for the tools.')
+@click.option('--local-scratch', type=click.PATH, required=False,
+              help='Temporary directory for storing intermediate files.')
+@click.option('--log-dir', type=click.PATH, required=False,
+              help='Directory containing log files.')
+@click.option('--restart', is_flag=True, default=False,
```
Comment (RNAer) [inline scripts/oecophylla.py:26]: "--force" is more conventional?

File: scripts/oecophylla.py [hunk 18]
```diff
@@ -4,11 +4,89 @@
...
+              help='Restarts the run and overwrites previous input.')
+@click.option('--output-dir',
+              help='Input directory of all of the samples.')
+# TODO: extra snakemake parameters with defaults
+
+# TODO: everything past here will be passed directly into snakemake
+def local_workflow():
+    """
```
Comment (RNAer) [inline scripts/oecophylla.py:34]: could you write a clearer description? this docstring is the help message exposed by `click` to end users

File: scripts/oecophylla.py [hunk 19]
```diff
@@ -4,11 +4,89 @@
...
+@click.option('--local-scratch',
+              help='Temporary directory for storing intermediate files.')
+@click.option('--qsub/--slurm', is_flag=True,
+              help='Temporary directory for storing intermediate files.')
+@click.option('--log-dir',
+              help='Directory containing log files.')
+@click.option('--output-dir',
+              help='Input directory of all of the samples.')
+def cluster_workflow():
+    """
```
Comment (RNAer) [inline scripts/oecophylla.py:65]: same here for doc

File: scripts/oecophylla.py [hunk 20]
```diff
@@ -4,11 +4,89 @@
...
+@click.option('--manifest',
+              help='Manifest file containing all of the paths (csv file).')
+@click.option('--params',
+              help='Specify parameters for the tools.')
+@click.option('--cluster-file',
+              help='Input biom table of abundances.')
+@click.option('--local-scratch',
+              help='Temporary directory for storing intermediate files.')
+@click.option('--qsub/--slurm', is_flag=True,
+              help='Temporary directory for storing intermediate files.')
```
Comment (RNAer) [inline scripts/oecophylla.py:59]: fix `help` here

File: scripts/oecophylla.py [hunk 21]
```diff
@@ -4,11 +4,89 @@
...
+
+@run.command()
+@click.option('--input-dir',
+              help='Input directory of all of the samples.')
+@click.option('--manifest',
+              help='Manifest file containing all of the paths (csv file).')
+@click.option('--params',
+              help='Specify parameters for the tools.')
+@click.option('--cluster-file',
+              help='Input biom table of abundances.')
```
Comment (RNAer) [inline scripts/oecophylla.py:55]: fix help. it is not biom table

**Review by RNAer — CHANGES_REQUESTED** [review]

File: oecophylla/util/parse.py (hunk 8 above)
Comment (wasade) [reply oecophylla/util/parse.py:37]: This was done in the inverse way to what I was thinking... will `sample_col` in `extract_sample_reads` ever not be `Sample`? Perhaps not worth blocking though

File: oecophylla/util/parse.py [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+       Delimiter for parsing the pandas input.
+
+    Returns
+    -------
+    data_df : pd.DataFrame
+       DataFrame containing the sample sheet information.
+
+    Note
+    ----
+    An example of a sample sheet would be as follows
```
Comment (wasade) [inline oecophylla/util/parse.py:174]: I'm left with a feeling of wanting more...

**Comment by mortonjt** [comment]
moving to dev

File: oecophylla/util/parse.py (hunk 22 above)
Comment (mortonjt) [reply oecophylla/util/parse.py:174]: That file can be found here
https://github.com/biocore/oecophylla/blob/master/test_data/example_sample_sheet.txt

**Pull request closed without merging**